File: app/Helpers/DiffRenderer.php
<?php

namespace App\Helpers;

/**
 * Custom GitHub-like diff renderer
 * Parses and renders git diffs in HTML format
 */
class DiffRenderer
{
    private object $diff;
    private array $files = [];
    private array $commits = [];

    public function __construct(object $diff)
    {
        $this->diff = $diff;
        $this->parseFiles();
        $this->parseCommits();
    }

    public function getFiles(): array
    {
        return $this->files;
    }

    public function getCommits(): array
    {
        return $this->commits;
    }

    private function parseCommits(): void
    {
    }

    /**
     * Parse files from GitHub compare API response
     * and transform patches into structured hunks/rows.
     */
    private function parseFiles(): void
    {
        $filesObject = $this->diff->files;
        $this->files = $filesObject;

        $parsed_files = [];

        foreach ($this->files as $file) {
            $file_to_parse = [];

            $file_to_parse['filename'] = $file->filename;
            $file_to_parse['status'] = $file->status;
            $file_to_parse['additions'] = $file->additions;
            $file_to_parse['deletions'] = $file->deletions;

            // Build diff hunks and aligned rows for this file
            $file_to_parse['changes'] = $this->formatPatchString($file->patch ?? null);

            $parsed_files[] = $file_to_parse;
        }

        $this->files = $parsed_files;
    }

    /**
     * Convert a unified diff patch string into structured hunks with
     * - old/new line blocks (raw)
     * - rows: aligned pairs for side-by-side display
     *   including intra-line segments for add/del pairs.
     */
    private function formatPatchString(?string $patch)
    {
        // Return a structure per the comment:
        // [
        //   [
        //     'old' => ['start' => int, 'end' => int, 'lines' => [[type, content], ...]],
        //     'new' => ['start' => int, 'end' => int, 'lines' => [[type, content], ...]],
        //   ],
        //   ...
        // ]

        $hunks = [];

        if ($patch === null || $patch === '') {
            return $hunks;
        }

        $lines = preg_split("/\r?\n/", $patch);
        if ($lines === false) {
            return $hunks;
        }

        $currentHunk = null;
        $oldLine = 0;
        $newLine = 0;

        $flushHunk = function () use (&$hunks, &$currentHunk, $oldLine, $newLine) {
            if ($currentHunk === null) {
                return;
            }
            // Set end positions based on last consumed line numbers
            $currentHunk['old']['end'] = $currentHunk['old']['start'] > 0
                ? max($currentHunk['old']['start'], $oldLine - 1)
                : 0;
            $currentHunk['new']['end'] = $currentHunk['new']['start'] > 0
                ? max($currentHunk['new']['start'], $newLine - 1)
                : 0;
            $hunks[] = $currentHunk;
            $currentHunk = null;
        };

        foreach ($lines as $rawLine) {
            // New hunk header
            if (preg_match('/^@@ -(?P<old_start>\d+)(?:,(?P<old_count>\d+))? \+(?P<new_start>\d+)(?:,(?P<new_count>\d+))? @@/', $rawLine, $m)) {
                // close the previous hunk if any
                $flushHunk();

                $oldStart = (int)($m['old_start'] ?? 0);
                $newStart = (int)($m['new_start'] ?? 0);
                $oldLine = $oldStart;
                $newLine = $newStart;

                $currentHunk = [
                    'old' => [
                        'start' => $oldStart,
                        'end' => $oldStart, // will be updated as we parse
                        'lines' => [],
                    ],
                    'new' => [
                        'start' => $newStart,
                        'end' => $newStart, // will be updated as we parse
                        'lines' => [],
                    ],
                ];
                continue;
            }

            // Outside any hunk; ignore non-hunk lines (file headers, metadata)
            if ($currentHunk === null) {
                continue;
            }

            // Handle special meta lines like "\\ No newline at end of file" — skip
            if (strlen($rawLine) > 0 && $rawLine[0] === '\\') {
                continue;
            }

            // Classify line by first char
            $prefix = $rawLine[0] ?? ' ';
            $content = ($prefix === '+' || $prefix === '-' || $prefix === ' ')
                ? substr($rawLine, 1)
                : $rawLine;

            switch ($prefix) {
                case '+':
                    // Addition: only affects new side
                    $currentHunk['new']['lines'][] = [
                        'type' => 'add',
                        'content' => $content,
                    ];
                    $newLine++;
                    break;

                case '-':
                    // Deletion: only affects old side
                    $currentHunk['old']['lines'][] = [
                        'type' => 'del',
                        'content' => $content,
                    ];
                    $oldLine++;
                    break;

                default:
                    // Context line: present on both sides
                    $currentHunk['old']['lines'][] = [
                        'type' => 'normal',
                        'content' => $content,
                    ];
                    $currentHunk['new']['lines'][] = [
                        'type' => 'normal',
                        'content' => $content,
                    ];
                    $oldLine++;
                    $newLine++;
                    break;
            }
        }

        // Flush the last hunk if present
        $flushHunk();

        // Post-process: combine old/new side lines into aligned rows per hunk
        foreach ($hunks as &$hunk) {
            $hunk['rows'] = $this->buildRowsForHunk($hunk);
        }
        unset($hunk);

        return $hunks;
    }

    /**
     * Given a hunk with old/new lines, generate aligned rows for side-by-side view.
     * Also computes intra-line segments for add/del pairs to enable fine-grained highlighting.
     */
    private function buildRowsForHunk(array $hunk): array
    {
        $oldLines = $hunk['old']['lines'] ?? [];
        $newLines = $hunk['new']['lines'] ?? [];

        // Build a mapping of which new line to pair with each old deletion
        $pairingMap = $this->buildPairingMap($oldLines, $newLines);

        $rows = [];
        $iOld = 0;
        $iNew = 0;
        $oldNo = $hunk['old']['start'] ?? null;
        $newNo = $hunk['new']['start'] ?? null;
        $usedNewIndices = [];

        while ($iOld < count($oldLines) || $iNew < count($newLines)) {
            $l = $iOld < count($oldLines) ? $oldLines[$iOld] : null;
            $r = $iNew < count($newLines) ? $newLines[$iNew] : null;

            // Context lines align 1:1
            if ($l && $r && ($l['type'] ?? '') === 'normal' && ($r['type'] ?? '') === 'normal') {
                $rows[] = [
                    'left' => ['number' => $oldNo, 'type' => 'normal', 'content' => $l['content'] ?? ''],
                    'right' => ['number' => $newNo, 'type' => 'normal', 'content' => $r['content'] ?? ''],
                ];
                $iOld++; $iNew++; $oldNo++; $newNo++;
                continue;
            }

            // Changed line: check if we have a pairing
            if ($l && ($l['type'] ?? '') === 'del' && isset($pairingMap[$iOld])) {
                $pairedNewIndex = $pairingMap[$iOld];

                // Output any standalone additions before this pairing
                while ($iNew < $pairedNewIndex) {
                    if (!isset($usedNewIndices[$iNew])) {
                        $newLine = $newLines[$iNew];
                        if (($newLine['type'] ?? '') === 'add') {
                            $rows[] = [
                                'left' => ['number' => null, 'type' => 'empty', 'content' => ''],
                                'right' => ['number' => $newNo, 'type' => 'add', 'content' => (string)($newLine['content'] ?? '')],
                            ];
                            $usedNewIndices[$iNew] = true;
                        }
                    }
                    $iNew++;
                    if ($newNo !== null) { $newNo++; }
                }

                // Now pair the deletion with the matched addition
                $matchedNewLine = $newLines[$pairedNewIndex];
                $leftContent = (string)($l['content'] ?? '');
                $rightContent = (string)($matchedNewLine['content'] ?? '');
                [$leftSeg, $rightSeg] = $this->computeIntralineSegments($leftContent, $rightContent);
                $whitespaceOnly = $this->isWhitespaceOnlyChange($leftContent, $rightContent);

                // Calculate the correct line number for the matched addition
                $matchedNewNo = $newNo;
                $tempI = $iNew;
                while ($tempI <= $pairedNewIndex) {
                    if ($tempI < $pairedNewIndex && !isset($usedNewIndices[$tempI])) {
                        $matchedNewNo++;
                    }
                    $tempI++;
                }

                $rows[] = [
                    'left' => [
                        'number' => $oldNo,
                        'type' => 'del',
                        'content' => $leftContent,
                        'segments' => $leftSeg,
                        'whitespace_only' => $whitespaceOnly,
                    ],
                    'right' => [
                        'number' => $matchedNewNo,
                        'type' => 'add',
                        'content' => $rightContent,
                        'segments' => $rightSeg,
                        'whitespace_only' => $whitespaceOnly,
                    ],
                ];
                $iOld++; $oldNo++;
                $usedNewIndices[$pairedNewIndex] = true;
                continue;
            }

            // Standalone deletion
            if ($l && ($l['type'] ?? '') === 'del') {
                $rows[] = [
                    'left' => ['number' => $oldNo, 'type' => 'del', 'content' => (string)($l['content'] ?? '')],
                    'right' => ['number' => null, 'type' => 'empty', 'content' => ''],
                ];
                $iOld++; $oldNo++;
                continue;
            }

            // Standalone addition
            if ($r && ($r['type'] ?? '') === 'add' && !isset($usedNewIndices[$iNew])) {
                $rows[] = [
                    'left' => ['number' => null, 'type' => 'empty', 'content' => ''],
                    'right' => ['number' => $newNo, 'type' => 'add', 'content' => (string)($r['content'] ?? '')],
                ];
                $iNew++; $newNo++;
                continue;
            }

            // Skip used indices
            if ($r && isset($usedNewIndices[$iNew])) {
                $iNew++;
                if ($newNo !== null) { $newNo++; }
                continue;
            }

            // Fallbacks: mirror remaining context if sides misaligned
            if ($l && ($l['type'] ?? '') === 'normal' && !$r) {
                $rows[] = [
                    'left' => ['number' => $oldNo, 'type' => 'normal', 'content' => (string)($l['content'] ?? '')],
                    'right' => ['number' => $newNo, 'type' => 'normal', 'content' => (string)($l['content'] ?? '')],
                ];
                $iOld++; $oldNo++; if ($newNo !== null) { $newNo++; }
                continue;
            }

            if ($r && ($r['type'] ?? '') === 'normal' && !$l && !isset($usedNewIndices[$iNew])) {
                $rows[] = [
                    'left' => ['number' => $oldNo, 'type' => 'normal', 'content' => (string)($r['content'] ?? '')],
                    'right' => ['number' => $newNo, 'type' => 'normal', 'content' => (string)($r['content'] ?? '')],
                ];
                $iNew++; $newNo++; if ($oldNo !== null) { $oldNo++; }
                continue;
            }

            // Safety: prevent infinite loop on unexpected input
            break;
        }

        return $rows;
    }

    /**
     * Build a mapping of old deletion indices to new addition indices based on whitespace-aware similarity.
     * Returns array where key is old index, value is the best matching new index.
     */
    private function buildPairingMap(array $oldLines, array $newLines): array
    {
        $pairingMap = [];
        $usedNewIndices = [];

        // Find all deletions and additions
        $deletions = [];
        $additions = [];

        foreach ($oldLines as $i => $line) {
            if (($line['type'] ?? '') === 'del') {
                $deletions[$i] = $line;
            }
        }

        foreach ($newLines as $i => $line) {
            if (($line['type'] ?? '') === 'add') {
                $additions[$i] = $line;
            }
        }

        // For each deletion, find the best matching addition
        foreach ($deletions as $oldIdx => $delLine) {
            $delContent = (string)($delLine['content'] ?? '');
            $delNoWs = preg_replace('/\s/u', '', $delContent);

            $bestScore = -1;
            $bestNewIdx = null;

            foreach ($additions as $newIdx => $addLine) {
                // Skip already used additions
                if (isset($usedNewIndices[$newIdx])) {
                    continue;
                }

                $addContent = (string)($addLine['content'] ?? '');
                $addNoWs = preg_replace('/\s/u', '', $addContent);

                // Perfect match ignoring whitespace
                if ($delNoWs === $addNoWs && $delNoWs !== '') {
                    $bestScore = 1.0;
                    $bestNewIdx = $newIdx;
                    break; // Take the first perfect match
                }

                // Partial similarity
                $similarity = $this->calculateSimilarity($delNoWs, $addNoWs);
                if ($similarity > $bestScore && $similarity > 0.7) {
                    $bestScore = $similarity;
                    $bestNewIdx = $newIdx;
                }
            }

            if ($bestNewIdx !== null) {
                $pairingMap[$oldIdx] = $bestNewIdx;
                $usedNewIndices[$bestNewIdx] = true;
            }
        }

        return $pairingMap;
    }

    /**
     * Check if two lines differ only in whitespace.
     * Returns true if the lines are identical when all whitespace is removed.
     */
    private function isWhitespaceOnlyChange(string $old, string $new): bool
    {
        $oldNoWs = preg_replace('/\s/u', '', $old);
        $newNoWs = preg_replace('/\s/u', '', $new);
        return $oldNoWs === $newNoWs && $old !== $new;
    }

    /**
     * Compute intra-line diff segments for a pair of texts.
     * Uses smart rules:
     * - Added side: Only highlight if removing the changes from new code gives old code
     * - Removed side: Only highlight if the line is 90%+ similar to new code
     */
    private function computeIntralineSegments(string $old, string $new): array
    {
        if ($old === $new) {
            $seg = [['text' => $old, 'type' => 'equal']];
            return [$seg, $seg];
        }

        // If only whitespace changed, return no highlights
        $oldTrimmed = trim($old);
        $newTrimmed = trim($new);
        if ($oldTrimmed === $newTrimmed) {
            return [
                [['text' => $old, 'type' => 'equal']],
                [['text' => $new, 'type' => 'equal']]
            ];
        }

        $oldLen = strlen($old);
        $newLen = strlen($new);

        // Longest common prefix
        $prefix = 0;
        $limit = min($oldLen, $newLen);
        while ($prefix < $limit && $old[$prefix] === $new[$prefix]) {
            $prefix++;
        }

        // Longest common suffix, ensuring no overlap with prefix
        $suffix = 0;
        while ($suffix < ($limit - $prefix)
            && $old[$oldLen - 1 - $suffix] === $new[$newLen - 1 - $suffix]) {
            $suffix++;
        }

        $oldMid = substr($old, $prefix, $oldLen - $prefix - $suffix);
        $newMid = substr($new, $prefix, $newLen - $prefix - $suffix);
        $pre = substr($old, 0, $prefix);
        $suf = $suffix > 0 ? substr($old, $oldLen - $suffix) : '';
        $preNew = substr($new, 0, $prefix);
        $sufNew = $suffix > 0 ? substr($new, $newLen - $suffix) : '';

        // Check if we should show highlights on the new (added) side
        // Only if removing the changed part from new gives us old
        $shouldHighlightNew = false;
        if ($newMid !== '') {
            $reconstructed = $preNew . $sufNew;
            if ($reconstructed === $old) {
                $shouldHighlightNew = true;
            }
        }

        // Check if we should show highlights on the old (removed) side
        // Only if the line is 90%+ similar after removing changes
        $shouldHighlightOld = false;
        if ($oldMid !== '') {
            $reconstructed = $pre . $suf;
            $similarity = $this->calculateSimilarity($reconstructed, $new);
            if ($similarity >= 0.9) {
                $shouldHighlightOld = true;
            }
        }

        $leftSeg = [];
        $rightSeg = [];

        if ($pre !== '') { $leftSeg[] = ['text' => $pre, 'type' => 'equal']; }
        if ($oldMid !== '' && $shouldHighlightOld) { $leftSeg[] = ['text' => $oldMid, 'type' => 'change']; }
        elseif ($oldMid !== '') { $leftSeg[] = ['text' => $oldMid, 'type' => 'equal']; }
        if ($suf !== '') { $leftSeg[] = ['text' => $suf, 'type' => 'equal']; }

        if ($preNew !== '') { $rightSeg[] = ['text' => $preNew, 'type' => 'equal']; }
        if ($newMid !== '' && $shouldHighlightNew) { $rightSeg[] = ['text' => $newMid, 'type' => 'change']; }
        elseif ($newMid !== '') { $rightSeg[] = ['text' => $newMid, 'type' => 'equal']; }
        if ($sufNew !== '') { $rightSeg[] = ['text' => $sufNew, 'type' => 'equal']; }

        // If either middle is empty, segments array may end up only equals — that's fine
        if (!$leftSeg) { $leftSeg[] = ['text' => $old, 'type' => 'equal']; }
        if (!$rightSeg) { $rightSeg[] = ['text' => $new, 'type' => 'equal']; }

        // If highlighted portion is 90%+ of the line, don't highlight
        $leftHighlightLen = $this->getSegmentsLength($leftSeg, 'change');
        $rightHighlightLen = $this->getSegmentsLength($rightSeg, 'change');

        if ($leftHighlightLen / max(1, $oldLen) >= 0.9) {
            $leftSeg = [['text' => $old, 'type' => 'equal']];
        }

        if ($rightHighlightLen / max(1, $newLen) >= 0.9) {
            $rightSeg = [['text' => $new, 'type' => 'equal']];
        }

        return [$leftSeg, $rightSeg];
    }

    /**
     * Get total length of segments with a specific type.
     */
    private function getSegmentsLength(array $segments, string $type): int
    {
        $length = 0;
        foreach ($segments as $seg) {
            if (($seg['type'] ?? '') === $type) {
                $length += strlen($seg['text'] ?? '');
            }
        }
        return $length;
    }

    /**
     * Calculate similarity between two strings using longest common subsequence.
     * Returns a value between 0 and 1.
     */
    private function calculateSimilarity(string $str1, string $str2): float
    {
        $len1 = strlen($str1);
        $len2 = strlen($str2);
        $maxLen = max($len1, $len2);

        if ($maxLen === 0) {
            return 1.0;
        }

        $lcs = $this->getLongestCommonSubsequenceLength($str1, $str2);
        return $lcs / $maxLen;
    }

    /**
     * Calculate length of longest common subsequence.
     */
    private function getLongestCommonSubsequenceLength(string $str1, string $str2): int
    {
        $len1 = strlen($str1);
        $len2 = strlen($str2);

        $dp = array_fill(0, $len1 + 1, array_fill(0, $len2 + 1, 0));

        for ($i = 1; $i <= $len1; $i++) {
            for ($j = 1; $j <= $len2; $j++) {
                if ($str1[$i - 1] === $str2[$j - 1]) {
                    $dp[$i][$j] = $dp[$i - 1][$j - 1] + 1;
                } else {
                    $dp[$i][$j] = max($dp[$i - 1][$j], $dp[$i][$j - 1]);
                }
            }
        }

        return $dp[$len1][$len2];
    }
}

File: app/services/diff_renderer.rb
# GitHub-like side-by-side diff renderer, ported from the Laravel app's
# DiffRenderer. Turns a GitHub compare/pull-files response into structured
# files -> hunks -> aligned left/right rows, pairing changed lines within a
# block (whitespace-insensitive + Levenshtein), dropping crossing pairs via a
# longest-increasing-subsequence, and computing intra-line change segments so
# only the changed span is highlighted.
#
# Output shape (all hashes, symbol keys):
#   files: [ { filename:, status:, additions:, deletions:, hunks: [ { rows: [
#     { left: cell, right: cell } ] } ] } ]
#   cell:  { number:, type: "normal"|"add"|"del"|"empty", content:,
#            segments: [ { text:, type: "equal"|"change" } ], whitespace_only: }
class DiffRenderer
  EXCLUDED_FILES = %w[package-lock.json composer.lock yarn.lock Gemfile.lock].freeze
  HUNK_HEADER = /\A@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@/

  # files_data is the "files" array from a compare / pull-files response.
  def initialize(files_data)
    @files_data = Array(files_data)
  end

  def files
    @files ||= @files_data.filter_map do |file|
      next if EXCLUDED_FILES.include?(File.basename(file["filename"].to_s))

      {
        filename: file["filename"],
        status: file["status"],
        additions: file["additions"],
        deletions: file["deletions"],
        hunks: format_patch(file["patch"])
      }
    end
  end

  private

  def format_patch(patch)
    return [] if patch.blank?

    hunks = []
    current = nil
    old_line = 0
    new_line = 0

    flush = lambda do
      next if current.nil?

      current[:old][:end] = current[:old][:start].positive? ? [ current[:old][:start], old_line - 1 ].max : 0
      current[:new][:end] = current[:new][:start].positive? ? [ current[:new][:start], new_line - 1 ].max : 0
      hunks << current
      current = nil
    end

    patch.split(/\r?\n/).each do |raw|
      if (m = raw.match(HUNK_HEADER))
        flush.call
        old_start = m[1].to_i
        new_start = m[3].to_i
        old_line = old_start
        new_line = new_start
        current = {
          old: { start: old_start, end: old_start, lines: [] },
          new: { start: new_start, end: new_start, lines: [] }
        }
        next
      end

      next if current.nil?
      next if raw.start_with?("\\") # "\ No newline at end of file"

      prefix = raw[0] || " "
      content = %w[+ - ].include?(prefix) || prefix == " " ? raw[1..] || "" : raw

      case prefix
      when "+"
        current[:new][:lines] << { type: "add", content: content }
        new_line += 1
      when "-"
        current[:old][:lines] << { type: "del", content: content }
        old_line += 1
      else
        current[:old][:lines] << { type: "normal", content: content }
        current[:new][:lines] << { type: "normal", content: content }
        old_line += 1
        new_line += 1
      end
    end

    flush.call

    hunks.each { |hunk| hunk[:rows] = build_rows(hunk) }
    hunks
  end

  def build_rows(hunk)
    old_lines = hunk[:old][:lines]
    new_lines = hunk[:new][:lines]
    pairing = build_pairing_map(old_lines, new_lines)

    rows = []
    i_old = 0
    i_new = 0
    old_no = hunk[:old][:start]
    new_no = hunk[:new][:start]
    used_new = {}

    while i_old < old_lines.size || i_new < new_lines.size
      l = old_lines[i_old]
      r = new_lines[i_new]

      # Context lines align 1:1.
      if l && r && l[:type] == "normal" && r[:type] == "normal"
        rows << {
          left: { number: old_no, type: "normal", content: l[:content] },
          right: { number: new_no, type: "normal", content: r[:content] }
        }
        i_old += 1; i_new += 1; old_no += 1; new_no += 1
        next
      end

      # Paired change: deletion matched to an addition.
      if l && l[:type] == "del" && pairing.key?(i_old)
        paired_new_index = pairing[i_old]

        # Emit standalone additions that come before the paired one.
        while i_new < paired_new_index
          unless used_new[i_new]
            nl = new_lines[i_new]
            if nl[:type] == "add"
              rows << {
                left: { number: nil, type: "empty", content: "" },
                right: { number: new_no, type: "add", content: nl[:content].to_s }
              }
              used_new[i_new] = true
            end
          end
          i_new += 1
          new_no += 1 unless new_no.nil?
        end

        matched = new_lines[paired_new_index]
        left_content = l[:content].to_s
        right_content = matched[:content].to_s
        left_seg, right_seg = compute_intraline_segments(left_content, right_content)
        whitespace_only = whitespace_only_change?(left_content, right_content)

        matched_new_no = new_no
        temp = i_new
        while temp <= paired_new_index
          matched_new_no += 1 if temp < paired_new_index && !used_new[temp]
          temp += 1
        end

        rows << {
          left: { number: old_no, type: "del", content: left_content, segments: left_seg, whitespace_only: whitespace_only },
          right: { number: matched_new_no, type: "add", content: right_content, segments: right_seg, whitespace_only: whitespace_only }
        }
        i_old += 1; old_no += 1
        used_new[paired_new_index] = true
        next
      end

      # Standalone deletion.
      if l && l[:type] == "del"
        rows << {
          left: { number: old_no, type: "del", content: l[:content].to_s },
          right: { number: nil, type: "empty", content: "" }
        }
        i_old += 1; old_no += 1
        next
      end

      # Standalone addition.
      if r && r[:type] == "add" && !used_new[i_new]
        rows << {
          left: { number: nil, type: "empty", content: "" },
          right: { number: new_no, type: "add", content: r[:content].to_s }
        }
        i_new += 1; new_no += 1
        next
      end

      # Skip an already-consumed addition.
      if r && used_new[i_new]
        i_new += 1
        new_no += 1 unless new_no.nil?
        next
      end

      # Fallbacks for misaligned sides.
      if l && l[:type] == "normal" && r.nil?
        rows << {
          left: { number: old_no, type: "normal", content: l[:content].to_s },
          right: { number: new_no, type: "normal", content: l[:content].to_s }
        }
        i_old += 1; old_no += 1
        new_no += 1 unless new_no.nil?
        next
      end

      if r && r[:type] == "normal" && l.nil? && !used_new[i_new]
        rows << {
          left: { number: old_no, type: "normal", content: r[:content].to_s },
          right: { number: new_no, type: "normal", content: r[:content].to_s }
        }
        i_new += 1; new_no += 1
        old_no += 1 unless old_no.nil?
        next
      end

      break # safety against unexpected input
    end

    rows
  end

  def build_pairing_map(old_lines, new_lines)
    pairing = {}
    used_new = {}

    old_blocks = assign_change_blocks(old_lines)
    new_blocks = assign_change_blocks(new_lines)

    deletions_by_block = Hash.new { |h, k| h[k] = {} }
    additions_by_block = Hash.new { |h, k| h[k] = {} }

    old_lines.each_with_index do |line, i|
      deletions_by_block[old_blocks[i]][i] = line if line[:type] == "del"
    end
    new_lines.each_with_index do |line, i|
      additions_by_block[new_blocks[i]][i] = line if line[:type] == "add"
    end

    deletions_by_block.each do |block, deletions|
      additions = additions_by_block[block] || {}

      deletions.each do |old_idx, del_line|
        del_content = del_line[:content].to_s
        del_no_ws = del_content.gsub(/\s/, "")

        best_perfect_raw = Float::INFINITY
        best_perfect_idx = nil
        best_partial = Float::INFINITY
        best_partial_idx = nil

        additions.each do |new_idx, add_line|
          next if used_new[new_idx]

          add_content = add_line[:content].to_s
          add_no_ws = add_content.gsub(/\s/, "")

          # Perfect match ignoring whitespace — prefer closest indentation.
          if del_no_ws == add_no_ws && !del_no_ws.empty?
            raw_distance = levenshtein(del_content, add_content)
            if raw_distance < best_perfect_raw
              best_perfect_raw = raw_distance
              best_perfect_idx = new_idx
              break if raw_distance.zero?
            end
            next
          end

          distance = levenshtein(del_no_ws, add_no_ws)
          avg_len = (del_no_ws.length + add_no_ws.length) / 2.0
          normalized = avg_len.positive? ? distance / avg_len : distance

          if normalized < best_partial && normalized < 0.3
            best_partial = normalized
            best_partial_idx = new_idx
          end
        end

        best_new_idx = best_perfect_idx || best_partial_idx
        if best_new_idx
          pairing[old_idx] = best_new_idx
          used_new[best_new_idx] = true
        end
      end
    end

    remove_crossing_pairs(pairing)
  end

  # Change lines between the Nth and (N+1)th context lines share block N, so
  # deletions only pair with additions in the same region.
  def assign_change_blocks(lines)
    block_map = {}
    block = 0
    lines.each_with_index do |line, i|
      if line[:type] == "normal"
        block += 1
      else
        block_map[i] = block
      end
    end
    block_map
  end

  # Keep only a non-crossing (monotonic) subset of pairs via LIS on new indices.
  def remove_crossing_pairs(pairing)
    return pairing if pairing.size <= 1

    keys = pairing.keys.sort
    values = keys.map { |k| pairing[k] }
    n = values.size

    dp = Array.new(n, 1)
    prev = Array.new(n, -1)
    best_len = 1
    best_end = 0

    (1...n).each do |i|
      (0...i).each do |j|
        if values[j] < values[i] && dp[i] < dp[j] + 1
          dp[i] = dp[j] + 1
          prev[i] = j
        end
      end
      if dp[i] > best_len
        best_len = dp[i]
        best_end = i
      end
    end

    positions = []
    pos = best_end
    while pos >= 0
      positions << pos
      pos = prev[pos]
    end
    positions.reverse!

    positions.each_with_object({}) { |p, filtered| filtered[keys[p]] = values[p] }
  end

  def whitespace_only_change?(old, new)
    old.gsub(/\s/, "") == new.gsub(/\s/, "") && old != new
  end

  # Intra-line segments: highlight only the changed span, with the same guards
  # as the Laravel version (reconstruction check on the new side, 90% LCS
  # similarity on the old side, and suppress highlights covering 90%+ of a line).
  def compute_intraline_segments(old, new)
    if old == new
      seg = [ { text: old, type: "equal" } ]
      return [ seg, seg ]
    end

    if old.strip == new.strip
      return [ [ { text: old, type: "equal" } ], [ { text: new, type: "equal" } ] ]
    end

    old_len = old.length
    new_len = new.length

    prefix = 0
    limit = [ old_len, new_len ].min
    prefix += 1 while prefix < limit && old[prefix] == new[prefix]

    suffix = 0
    while suffix < (limit - prefix) && old[old_len - 1 - suffix] == new[new_len - 1 - suffix]
      suffix += 1
    end

    old_mid = old[prefix, old_len - prefix - suffix]
    new_mid = new[prefix, new_len - prefix - suffix]
    pre = old[0, prefix]
    suf = suffix.positive? ? old[old_len - suffix..] : ""
    pre_new = new[0, prefix]
    suf_new = suffix.positive? ? new[new_len - suffix..] : ""

    should_highlight_new = !new_mid.empty? && (pre_new + suf_new) == old
    should_highlight_old = !old_mid.empty? && calculate_similarity(pre + suf, new) >= 0.9

    left_seg = []
    left_seg << { text: pre, type: "equal" } unless pre.empty?
    if !old_mid.empty? && should_highlight_old
      left_seg << { text: old_mid, type: "change" }
    elsif !old_mid.empty?
      left_seg << { text: old_mid, type: "equal" }
    end
    left_seg << { text: suf, type: "equal" } unless suf.empty?

    right_seg = []
    right_seg << { text: pre_new, type: "equal" } unless pre_new.empty?
    if !new_mid.empty? && should_highlight_new
      right_seg << { text: new_mid, type: "change" }
    elsif !new_mid.empty?
      right_seg << { text: new_mid, type: "equal" }
    end
    right_seg << { text: suf_new, type: "equal" } unless suf_new.empty?

    left_seg << { text: old, type: "equal" } if left_seg.empty?
    right_seg << { text: new, type: "equal" } if right_seg.empty?

    left_seg = [ { text: old, type: "equal" } ] if segments_length(left_seg, "change") / [ 1, old_len ].max.to_f >= 0.9
    right_seg = [ { text: new, type: "equal" } ] if segments_length(right_seg, "change") / [ 1, new_len ].max.to_f >= 0.9

    [ left_seg, right_seg ]
  end

  def segments_length(segments, type)
    segments.select { |s| s[:type] == type }.sum { |s| s[:text].length }
  end

  def calculate_similarity(str1, str2)
    max_len = [ str1.length, str2.length ].max
    return 1.0 if max_len.zero?

    lcs_length(str1, str2) / max_len.to_f
  end

  def lcs_length(str1, str2)
    len1 = str1.length
    len2 = str2.length
    dp = Array.new(len1 + 1) { Array.new(len2 + 1, 0) }

    (1..len1).each do |i|
      (1..len2).each do |j|
        dp[i][j] = if str1[i - 1] == str2[j - 1]
          dp[i - 1][j - 1] + 1
        else
          [ dp[i - 1][j], dp[i][j - 1] ].max
        end
      end
    end

    dp[len1][len2]
  end

  def levenshtein(a, b)
    DidYouMean::Levenshtein.distance(a, b)
  end
end
